src/Compilers/Server/ServerShared/ServerClient.cs

```diff
@@ -0,0 +1,159 @@
+{
+    internal abstract class ServerClient
+    {
+        internal int Run(string[] args)
+        {
+            string pipeName;
+            bool shutdown;
+            if (!ParseCommandLine(args, out pipeName, out shutdown))
+            {
+                return CommonCompiler.Failed;
+            }
+
+            pipeName = pipeName ?? GetDefaultPipeName();
+            var cancellationTokenSource = new CancellationTokenSource();
+            Console.CancelKeyPress += (sender, e) => { cancellationTokenSource.Cancel(); };
+
+            return shutdown
+                ? RunShutdown(pipeName, cancellationToken: cancellationTokenSource.Token)
+                : RunServer(pipeName, cancellationToken: cancellationTokenSource.Token);
+        }
+
+        protected abstract TimeSpan? GetKeepAliveTimeout();
+
+        protected abstract string GetDefaultPipeName();
+
+        protected abstract IClientConnectionHost CreateClientConnectionHost(string pipeName);
+
+        protected abstract Task<Stream> ConnectForShutdownAsync(string pipeName, int timeout);
+
+        /// <summary>
+        /// Was a server running with the specified session key during the execution of this call?
+        /// </summary>
+        protected virtual bool? WasServerRunning(string pipeName)
+        {
+            return null;
+        }
+
+        protected virtual int RunServerCore(string pipeName, IClientConnectionHost connectionHost, IDiagnosticListener listener, TimeSpan? keepAlive, CancellationToken cancellationToken)
+        {
+            CompilerServerLogger.Log("Keep alive timeout is: {0} milliseconds.", keepAlive?.TotalMilliseconds ?? 0);
+            FatalError.Handler = FailFast.OnFatalException;
+
+            var dispatcher = new ServerDispatcher(connectionHost, listener);
+            dispatcher.ListenAndDispatchConnections(keepAlive, cancellationToken);
+            return CommonCompiler.Succeeded;
+        }
+
+        internal int RunServer(string pipeName, IClientConnectionHost clientConnectionHost = null, IDiagnosticListener listener = null, TimeSpan? keepAlive = null, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            keepAlive = keepAlive ?? GetKeepAliveTimeout();
+            listener = listener ?? new EmptyDiagnosticListener();
+            clientConnectionHost = clientConnectionHost ?? CreateClientConnectionHost(pipeName);
+            return RunServerCore(pipeName, clientConnectionHost, listener, keepAlive, cancellationToken);
+        }
+
+        internal int RunShutdown(string pipeName, bool waitForProcess = true, TimeSpan? timeout = null, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return RunShutdownAsync(pipeName, waitForProcess, timeout, cancellationToken).GetAwaiter().GetResult();
+        }
+
+        /// <summary>
+        /// Shutting down the server is an inherently racy operation.  The server can be started or stopped by
```

Comment:
We resolved this by using a `Mutex` in the desktop version. Is there no equivalent portable implementation?